File: packages/vault_native/android/src/main/kotlin/dev/localhold/localhold_key_bridge/RecoveryMnemonicCodec.kt
// SPDX-License-Identifier: MPL-2.0
package dev.localhold.localhold_key_bridge

import android.content.Context
import java.security.MessageDigest
import java.util.Locale

/** BIP-39 entropy/checksum/index encoding only; wallet seed derivation is absent. */
internal class RecoveryMnemonicCodec internal constructor(candidateWords: List<String>) {
    internal constructor(context: Context) : this(
        context.assets.open(ASSET_NAME).bufferedReader(Charsets.UTF_8)
            .useLines { lines -> lines.filter(String::isNotBlank).toList() },
    )

    private val words: List<String> = candidateWords.toList().also(::validateWordList)
    private val indices = words.withIndex().associate { it.value to it.index }

    fun encode(entropy: ByteArray): List<String> {
        require(entropy.size == ENTROPY_BYTES)
        val checksum = MessageDigest.getInstance("SHA-256").digest(entropy)[0].toInt() and 0xff
        return List(WORD_COUNT) { wordPosition ->
            var index = 0
            repeat(BITS_PER_WORD) { bitWithinWord ->
                val bitPosition = wordPosition * BITS_PER_WORD + bitWithinWord
                val bit = if (bitPosition < ENTROPY_BITS) {
                    val source = entropy[bitPosition / 8].toInt() and 0xff
                    (source shr (7 - bitPosition % 8)) and 1
                } else {
                    (checksum shr (7 - (bitPosition - ENTROPY_BITS))) and 1
                }
                index = (index shl 1) or bit
            }
            words[index]
        }
    }

    fun decode(input: List<String>): ByteArray {
        if (input.size != WORD_COUNT) throw RecoveryMnemonicFailure()
        val normalized = input.map { it.lowercase(Locale.US) }
        val entropy = ByteArray(ENTROPY_BYTES)
        var encodedChecksum = 0
        normalized.forEachIndexed { wordPosition, word ->
            val index = indices[word] ?: throw RecoveryMnemonicFailure()
            repeat(BITS_PER_WORD) { bitWithinWord ->
                val bitPosition = wordPosition * BITS_PER_WORD + bitWithinWord
                val bit = (index shr (BITS_PER_WORD - 1 - bitWithinWord)) and 1
                if (bitPosition < ENTROPY_BITS) {
                    val target = bitPosition / 8
                    entropy[target] = (
                        (entropy[target].toInt() and 0xff) or
                            (bit shl (7 - bitPosition % 8))
                        ).toByte()
                } else {
                    encodedChecksum = (encodedChecksum shl 1) or bit
                }
            }
        }
        val expected = MessageDigest.getInstance("SHA-256").digest(entropy)[0].toInt() and 0xff
        if (encodedChecksum != expected) {
            entropy.fill(0)
            throw RecoveryMnemonicFailure()
        }
        return entropy
    }

    private fun validateWordList(candidate: List<String>) {
        require(candidate.size == 2048)
        require(candidate.toSet().size == candidate.size)
        require(candidate == candidate.sorted())
        require(candidate.all { it.matches(Regex("^[a-z]+$")) })
    }

    private companion object {
        const val ASSET_NAME = "bip39_english.txt"
        const val ENTROPY_BYTES = 32
        const val ENTROPY_BITS = ENTROPY_BYTES * 8
        const val WORD_COUNT = 24
        const val BITS_PER_WORD = 11
    }
}

internal class RecoveryMnemonicFailure : RuntimeException()

File: packages/vault_native/ios/localhold_key_bridge/Sources/localhold_key_bridge/RecoveryMnemonicCodec.swift
// SPDX-License-Identifier: MPL-2.0

import CryptoKit
import Foundation

/// BIP-39 entropy/checksum/index encoding only; wallet seed derivation is absent.
final class RecoveryMnemonicCodec {
  private let words: [String]
  private let indices: [String: Int]

  convenience init() throws {
    guard let url = Self.wordListURL(),
          let source = try? String(contentsOf: url, encoding: .utf8)
    else { throw RecoveryMnemonicFailure.invalidWordList }
    let candidate = source.split(whereSeparator: \Character.isWhitespace).map(String.init)
    try self.init(words: candidate)
  }

  init(words candidate: [String]) throws {
    guard candidate.count == 2048,
          Set(candidate).count == candidate.count,
          candidate == candidate.sorted(),
          candidate.allSatisfy({ $0.range(of: "^[a-z]+$", options: .regularExpression) != nil })
    else { throw RecoveryMnemonicFailure.invalidWordList }
    words = candidate
    indices = Dictionary(uniqueKeysWithValues: candidate.enumerated().map { ($0.element, $0.offset) })
  }

  func encode(_ entropy: Data) throws -> [String] {
    guard entropy.count == 32 else { throw RecoveryMnemonicFailure.invalidEntropy }
    let checksum = Self.checksumByte(for: entropy)
    return try (0..<24).map { wordPosition in
      var index = 0
      for bitWithinWord in 0..<11 {
        let bitPosition = wordPosition * 11 + bitWithinWord
        let bit: Int
        if bitPosition < 256 {
          bit = Int((entropy[bitPosition / 8] >> UInt8(7 - bitPosition % 8)) & 1)
        } else {
          bit = Int((checksum >> UInt8(7 - (bitPosition - 256))) & 1)
        }
        index = (index << 1) | bit
      }
      guard words.indices.contains(index) else { throw RecoveryMnemonicFailure.invalidEntropy }
      return words[index]
    }
  }

  func decode(_ input: [String]) throws -> Data {
    guard input.count == 24 else { throw RecoveryMnemonicFailure.invalidPhrase }
    var entropy = Data(repeating: 0, count: 32)
    var encodedChecksum = 0
    for (wordPosition, originalWord) in input.enumerated() {
      let word = originalWord.lowercased(with: Locale(identifier: "en_US_POSIX"))
      guard let index = indices[word] else { throw RecoveryMnemonicFailure.invalidPhrase }
      for bitWithinWord in 0..<11 {
        let bitPosition = wordPosition * 11 + bitWithinWord
        let bit = (index >> (10 - bitWithinWord)) & 1
        if bitPosition < 256 {
          let target = bitPosition / 8
          entropy[target] |= UInt8(bit << (7 - bitPosition % 8))
        } else {
          encodedChecksum = (encodedChecksum << 1) | bit
        }
      }
    }
    guard encodedChecksum == Int(Self.checksumByte(for: entropy)) else {
      entropy.resetBytes(in: 0..<entropy.count)
      throw RecoveryMnemonicFailure.invalidPhrase
    }
    return entropy
  }

  private static func checksumByte(for entropy: Data) -> UInt8 {
    Data(SHA256.hash(data: entropy))[0]
  }

  private static func wordListURL() -> URL? {
#if SWIFT_PACKAGE
    return Bundle.module.url(forResource: "bip39_english", withExtension: "txt")
#else
    let frameworkBundle = Bundle(for: RecoveryMnemonicCodec.self)
    if let resourceBundleURL = frameworkBundle.url(
      forResource: "localhold_key_bridge_recovery",
      withExtension: "bundle"
    ), let resourceBundle = Bundle(url: resourceBundleURL) {
      return resourceBundle.url(forResource: "bip39_english", withExtension: "txt")
    }
    return frameworkBundle.url(forResource: "bip39_english", withExtension: "txt")
#endif
  }
}

enum RecoveryMnemonicFailure: Error {
  case invalidWordList
  case invalidEntropy
  case invalidPhrase
}
